src/modules/services/keycloak.nix

```diff
@@ -0,0 +1,493 @@
+{ config
+, lib
+, pkgs
+, ...
+}:
+
+let
+  cfg = config.services.keycloak;
+
+  inherit (lib)
+    mkIf
+    mkMerge
+    mkOption
+    mkRenamedOptionModule
+    mkPackageOption
+    types
+    ;
+
+  inherit (types)
+    nullOr
+    oneOf
+    listOf
+    attrsOf
+    ;
+
+  assertStringPath =
+    optionName: value:
+    if builtins.isPath value then
+      throw ''
+        services.keycloak.${optionName}:
+          ${builtins.toString value}
+          is a Nix path, but should be a string, since Nix
+          paths are copied into the world-readable Nix store.
+      ''
+    else
+      value;
+in
+{
+  imports = [
+    (mkRenamedOptionModule [ "keycloak" "enable" ] [ "services" "keycloak" "enable" ])
+  ];
+
+  options.services.keycloak = {
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      example = true;
+      description = ''
+        Whether to enable the Keycloak identity and access management
+        server.
+      '';
+    };
+
+    sslCertificate = mkOption {
+      type = nullOr types.path;
+      default = null;
+      example = "/run/keys/ssl_cert";
+      apply = assertStringPath "sslCertificate";
+      description = ''
+        The path to a PEM formatted certificate to use for TLS/SSL
+        connections.
+      '';
+    };
+
+    sslCertificateKey = mkOption {
+      type = nullOr types.path;
+      default = null;
+      example = "/run/keys/ssl_key";
+      apply = assertStringPath "sslCertificateKey";
+      description = ''
+        The path to a PEM formatted private key to use for TLS/SSL
+        connections.
+      '';
+    };
+
+    plugins = mkOption {
+      type = listOf types.path;
+      default = [ ];
+      description = ''
+        Keycloak plugin jar, ear files or derivations containing
+        them. Packaged plugins are available through
+        `pkgs.keycloak.plugins`.
+      '';
+    };
+
+    database = {
+      type = mkOption {
+        type = types.enum [
```

Comment:
Since it would also work with postgres or mariadb does this enum really make sense? By default it's anyway using a file-based database which should be stored in the devenv state path.

Reply:
Yes that includes secrets and everything. Thats exactly what we want as well, so everything is deterministic if you reimport it.

Reply:
so far we only support `dev-mem` and `dev-file` thats why I only added those, the otherones need more complex setup etc etc. Does that make sense?